fitbenchmarking/parsing/fitbenchmark_parser.py

```diff
@@ -235,101 +236,166 @@ def _get_data_problem_entries(self):
+          - int
+          - float
+          - strings (not containing '[](),=' )
+
+        :param func: The definition to parse
+        :type func: str
+        :raises ParsingError: If unexpected characters are encoutered or
+                              parentheses do not match
+        :return: The function as a dict of name, value pairs.
+        :rtype: dict
+        """
+        # pylint: disable=too-many-branches
+        lhs, rhs = func.strip().split('=', 1)
+        name = lhs
+        if not re.match(r'^[A-Za-z0-9_]+$', name):
+            raise ParsingError(
+                f'Unexpected character in parameter name: {name}')
+
+        if rhs[0] in '([':
+            count = 0
+            if rhs[0] == '[':
+                delim = '[]'
+            else:  # '('
+                delim = '()'
+
+            for i, c in enumerate(rhs):
+                if c == delim[0]:
+                    count += 1
+                elif c == delim[1]:
+                    count -= 1
+                if count == 0:
+                    value = rhs[:i]
+                    rem = rhs[i+1:].strip(',')
+                    break
+            else:
+                raise ParsingError('Not all brackets are closed in function.')
+
+            if delim == '()':
+                value = self._parse_single_function(value[1:].strip())
+            else:  # []
+                value = [self._parse_function_value(v.strip())
+                         for v in value[1:].split(',') if v != '']
+        else:
+            if ',' in rhs:
+                value, rem = rhs.split(',', 1)
+            else:
+                value = rhs
+                rem = ''
+            value = self._parse_function_value(value)
+
+        func_dict = {name: value}
+        if rem:
+            func_dict.update(self._parse_single_function(rem))
+        return func_dict
+
+    @staticmethod
+    def _parse_function_value(value: str):
+        """
+        Parse a value from a string into a numerical type if possible.
+
+        :param value: The value to parse
+        :type value: str
+        :raises ParsingError: If the value has any unexpected characters
+        :return: The parsed value
+        :rtype: int, float, or str
+        """
+        if value.lower() == 'true':
+            return True
+        if value.lower() == 'false':
+            return False
+        for convert in [int, float, str]:
+            try:
+                return convert(value)
+            except (ValueError, TypeError):
+                continue
+        return value
+
+    @staticmethod
+    def _get_data_points(data_file_path):
+        """
+        Get the data points of the problem from the data file.
+
+        :param data_file_path: The path to the file to load the points from
+        :type data_file_path: str
+
+        :return: data
+        :rtype: dict<str, np.ndarray>
+        """
+
+        with open(data_file_path, 'r') as f:
+            data_text = f.readlines()
```

Comment:
I don't think this should be refactored as part of this pr.

I think this only came up as it was reordered so that methods more likely to be overridden are nearer the top of the file.

I will create an issue from the comment though as the numpy one is probably what we want.